testsuite/src/main/java/io/netty/testsuite/transport/socket/SocketChannelNotYetConnectedTest.java

```diff
@@ -68,4 +79,44 @@ private static void checkThrowable(Throwable cause) throws Throwable {
             throw cause;
         }
     }
+
+    @Test
+    @Timeout(30)
+    public void readMustBePendingUntilChannelIsActive(TestInfo info) throws Throwable {
+        run(info, new Runner<Bootstrap>() {
+            @Override
+            public void run(Bootstrap bootstrap) throws Throwable {
+                NioEventLoopGroup group = new NioEventLoopGroup();
+                ServerBootstrap sb = new ServerBootstrap().group(group);
+                Channel serverChannel = sb.childHandler(new ChannelInboundHandlerAdapter() {
+                    @Override
+                    public void channelActive(ChannelHandlerContext ctx) throws Exception {
+                        ctx.writeAndFlush(Unpooled.copyInt(42));
+                    }
+                }).channel(NioServerSocketChannel.class).bind(0).sync().channel();
+
+                final CountDownLatch readLatch = new CountDownLatch(1);
+                bootstrap.handler(new ByteToMessageDecoder() {
+                    @Override
+                    public void handlerAdded(ChannelHandlerContext ctx) throws Exception {
+                        assertFalse(ctx.channel().isActive());
+                        ctx.read();
+                    }
+
+                    @Override
+                    protected void decode(ChannelHandlerContext ctx, ByteBuf in, List<Object> out) throws Exception {
+                        assertThat(in.readableBytes()).isLessThanOrEqualTo(Integer.BYTES);
+                        if (in.readableBytes() == Integer.BYTES) {
+                            assertThat(in.readInt()).isEqualTo(42);
+                            readLatch.countDown();
+                        }
+                    }
+                });
+                bootstrap.connect(serverChannel.localAddress());
```

Comment:
maybe also call `sync()` here just in case that the connect fails. So we have some more details.